sssd_test_framework/utils/sbus.py

```diff
@@ -0,0 +1,672 @@
+
+    def _find_interface(self, elem: xml.Element) -> xml.Element | None:
+        """
+        Find the interface the provided XML element belongs to.
+        """
+        for child in self._xml:
+            if child.tag == "interface":
+                found = self._find_child(child, elem.attrib["name"])
+                if found == elem:
+                    return child
+
+        return None
+
+    def _make_prop(self, child: xml.Element) -> ProxyProperty:
+        """
+        Add a D-Bus proxy property to the current node.
+        """
+        iface = self._find_interface(child)
+        if iface is None:
+            raise RuntimeError(f"No interface found for {child.attrib['name']}")
+
+        prop = self.ProxyProperty(
+            self.host, self._dest, self._objpath, self._bus, interface=iface.attrib["name"], child=child
+        )
+        self._properties[child.attrib["name"]] = prop
+        return prop.get_value()
+
+    def _make_method(self, child) -> ProxyMethod:
+        """
+        Add a D-Bus proxy method to the current node.
+        """
+        iface = self._find_interface(child)
+        if iface is None:
+            raise RuntimeError(f"No interface found for {child.attrib['name']}")
+
+        method = self.ProxyMethod(
+            self.host, self._dest, self._objpath, self._bus, interface=iface.attrib["name"], child=child
+        )
+        self.__setattr__(child.attrib["name"], method)
+        return method
+
+    def _make_node(self, child) -> DBUSObject:
+        """
+        Add a D-Bus child node to the current node.
+        """
+        node = DBUSObject(
+            self.host, objpath=self._objpath + "/" + child.attrib["name"], dest=self._dest, bus=self._bus
+        )
+        self.__setattr__(child.attrib["name"], node)
+        return node
+
+    def _make_attr(self, child):
+        """
+        Make a D-Bus attribute and add it to the current node.
+        """
+        match child.tag:
+            case "property":
+                obj = self._make_prop(child)
+            case "method":
+                obj = self._make_method(child)
+            case "node":
+                obj = self._make_node(child)
+            case _:
+                raise AttributeError(
+                    f"'{self.__class__.__name__}' object has an attribute '{child.attrib['name']}'"
+                    f" of unkown type '{child.tag}'"
+                )
+
+        return obj
+
+    def __getattr__(self, attr) -> Any:
+        """
+        When an acccessing an attribute that doesn't belong to the object,
+        create it if it should exist.
+        """
+
+        # Required for the initialization
+        if attr == "_properties":
+            return {}
+
+        if attr in self._properties:
+            return self._properties[attr].get_value()
+
+        c = self._find_child(self._xml, attr)
+        if c is not None:
+            return self._make_attr(c)
+
+        return super().__getattribute__(attr)
```

Comment:
Why? You already know that the attribute does not exist so calling __getattribute__ will either raise AttributeError or go into infinite recursion, not sure which. You want to raise AttributeError here.

Reply:
The parent class could have the attribute. This is the recommended behavior.